### PR TITLE
172910 - get permit keys from pks stub

### PR DESCRIPTION
**Pull request by MastekRohitChettri**
This PBI consists changes of [PBI 172910](https://dev.azure.com/ukhydro/S-100%20Permit%20Generator/_workitems/edit/172910)

- Code changes added
- Unit tests added
- Functional test cases added

### CODE REVIEW

**Review by Nadeem14456 — CHANGES_REQUESTED** [review]
Hi @MastekRohitChettri ,

ProductkeyService does not adhere to Microsoft’s coding standards.

Please update this and ensure that all other places in the codebase follow the PascalCase convention.

**Comment by MastekRohitChettri** [comment]
> Hi @MastekRohitChettri ,
> 
> ProductkeyService does not adhere to Microsoft’s coding standards.
> 
> Please update this and ensure that all other places in the codebase follow the PascalCase convention.

Hi @Nadeem14456  ProductkeyService is PascalCase convention, can you suggest right name to replace

File: tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductkeyServiceApiClientTests.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        }
+
+        [Test]
+        [TestCase(HttpStatusCode.BadRequest)]
+        [TestCase(HttpStatusCode.NotFound)]
+        [TestCase(HttpStatusCode.Unauthorized)]
+        [TestCase(HttpStatusCode.InternalServerError)]
+        [TestCase(HttpStatusCode.ServiceUnavailable)]
+        [TestCase(HttpStatusCode.UnsupportedMediaType)]
+        public void WhenProductKeyServiceResponseOtherThanOk(HttpStatusCode httpStatusCode)
```
Comment (shirinbt14926) [inline tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductkeyServiceApiClientTests.cs]:
```suggestion
        public void WhenProductKeyServiceResponseOtherThanOk_ThenResponseShouldNotBeOk(HttpStatusCode httpStatusCode)
```

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/ProductkeyServiceTests.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+                && call.GetArgument<LogLevel>(0) == LogLevel.Information
+                && call.GetArgument<EventId>(1) == EventIds.ProductKeyServicePostPermitKeyRequestCompleted.ToEventId()
+                && call.GetArgument<IEnumerable<KeyValuePair<string, object>>>(2)!.ToDictionary(c => c.Key, c => c.Value)["{OriginalFormat}"].ToString() == "Request to ProductKeyService POST Uri : {RequestUri} completed. | StatusCode : {StatusCode}"
+            ).MustHaveHappenedOnceExactly();
+        }
+
+        [Test]
+        [TestCase(HttpStatusCode.BadRequest)]
+        [TestCase(HttpStatusCode.NotFound)]
+        public async Task WhenRequestsInvalidOrNonExistentData_ThenThrowException(HttpStatusCode httpStatusCode)
```
Comment (shirinbt14926) [inline tests/UKHO.S100PermitService.Common.UnitTests/Services/ProductkeyServiceTests.cs]:
```suggestion
        public async Task WhenRequestIsInvalidOrNonExistData_ThenThrowException(HttpStatusCode httpStatusCode)
```

File: src/UKHO.S100PermitService.API/Program.cs [hunk 3]
```diff
@@ -114,8 +113,9 @@ private static void ConfigureServices(WebApplicationBuilder builder)
             builder.Services.AddDistributedMemoryCache();
 
             builder.Services.Configure<EventHubLoggingConfiguration>(builder.Configuration.GetSection("EventHubLoggingConfiguration"));
-            builder.Services.Configure<HoldingsServiceApiConfiguration>(configuration.GetSection(HoldingsServiceApiConfiguration));
-            builder.Services.Configure<UserPermitServiceApiConfiguration>(configuration.GetSection(UserPermitServiceApiConfiguration));
+            builder.Services.Configure<HoldingsServiceApiConfiguration>(configuration.GetSection("HoldingsServiceApiConfiguration"));
+            builder.Services.Configure<UserPermitServiceApiConfiguration>(configuration.GetSection("UserPermitServiceApiConfiguration"));
+            builder.Services.Configure<ProductKeyServiceApiConfiguration>(builder.Configuration.GetSection("ProductKeyServiceApiConfiguration"));
```
Comment (shirinbt14926) [inline src/UKHO.S100PermitService.API/Program.cs:118]: To maintan consistent can we configure as
```suggestion
            builder.Services.Configure<ProductKeyServiceApiConfiguration>(configuration.GetSection("ProductKeyServiceApiConfiguration"));
```

**Review by shirinbt14926 — CHANGES_REQUESTED** [review]
Few minor changes to address please

File: src/UKHO.S100PermitService.Common/Clients/ProductkeyServiceApiClient.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        public async Task<HttpResponseMessage> CallProductkeyServiceApiAsync(string uri, HttpMethod httpMethod, string payload, string accessToken, string correlationId)
+        {
+            using var httpRequestMessage = new HttpRequestMessage(httpMethod, uri);
+            httpRequestMessage.Content = new StringContent(payload, Encoding.UTF8, "application/json");
+
+            if(!string.IsNullOrEmpty(accessToken))
+            {
+                httpRequestMessage.SetBearerToken(accessToken);
+                httpRequestMessage.AddHeader("X-Correlation-ID", correlationId);
+            }
```
[the remark this answers is not in the document]
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.Common/Clients/ProductkeyServiceApiClient.cs]: Resolved

File: tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductkeyServiceApiClientTests.cs (hunk 1 above)
Comment (MastekRohitChettri) [reply tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductkeyServiceApiClientTests.cs]: Resolved

File: tests/UKHO.S100PermitService.Common.UnitTests/Services/ProductkeyServiceTests.cs (hunk 2 above)
Comment (MastekRohitChettri) [reply tests/UKHO.S100PermitService.Common.UnitTests/Services/ProductkeyServiceTests.cs]: Resolved

File: src/UKHO.S100PermitService.API/Program.cs (hunk 3 above)
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.API/Program.cs]: Resolved

File: tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductKeyServiceApiClientTests.cs [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        private IProductKeyServiceApiClient? _productKeyServiceApiClient;
+
+        [SetUp]
+        public void SetUp()
+        {
+            _fakeHttpClientFactory = A.Fake<IHttpClientFactory>();
+        }
+
+        [Test]
+        public void WhenValidDataIsPassed_ThenProductKeyServiceReturnsOKResponse()
```
Comment (harshal11869) [inline tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductKeyServiceApiClientTests.cs:26]:
```suggestion
        public void WhenValidDataIsPassed_ThenProductKeyServiceReturnsOkResponse()
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductKeyServiceApiClientTests.cs (hunk 5 above)
Comment (MastekRohitChettri) [reply tests/UKHO.S100PermitService.Common.UnitTests/Helpers/ProductKeyServiceApiClientTests.cs]: Resolved

File: src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+        public async Task<HttpResponseMessage> CallProductKeyServiceApiAsync(string uri, HttpMethod httpMethod, string payLoad, string accessToken, string correlationId)
+        {
+            using var httpRequestMessage = new HttpRequestMessage(httpMethod, uri);
+            httpRequestMessage.Content = new StringContent(payLoad, Encoding.UTF8, "application/json");
+
+            if(!string.IsNullOrEmpty(accessToken))
+            {
+                httpRequestMessage.SetBearerToken(accessToken);
+                httpRequestMessage.AddHeader("X-Correlation-ID", correlationId);
```
Comment (harshal11869) [inline src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs:22]: Please use PermitServiceConstants here (XCorrelationIdHeaderKey)

**Review by harshal11869 — CHANGES_REQUESTED** [review]

**Review by shirinbt14926 — APPROVED** [review]
LGTM. Thanks

File: src/UKHO.S100PermitService.API/Program.cs [hunk 7]
```diff
@@ -113,9 +112,10 @@ private static void ConfigureServices(WebApplicationBuilder builder)
             builder.Services.AddEndpointsApiExplorer();
             builder.Services.AddDistributedMemoryCache();
 
-            builder.Services.Configure<EventHubLoggingConfiguration>(builder.Configuration.GetSection("EventHubLoggingConfiguration"));
-            builder.Services.Configure<HoldingsServiceApiConfiguration>(configuration.GetSection(HoldingsServiceApiConfiguration));
-            builder.Services.Configure<UserPermitServiceApiConfiguration>(configuration.GetSection(UserPermitServiceApiConfiguration));
+            builder.Services.Configure<EventHubLoggingConfiguration>(configuration.GetSection("EventHubLoggingConfiguration"));
+            builder.Services.Configure<HoldingsServiceApiConfiguration>(configuration.GetSection("HoldingsServiceApiConfiguration"));
+            builder.Services.Configure<UserPermitServiceApiConfiguration>(configuration.GetSection("UserPermitServiceApiConfiguration"));
+            builder.Services.Configure<ProductKeyServiceApiConfiguration>(configuration.GetSection("ProductKeyServiceApiConfiguration"));
```
Comment (jrippington) [inline src/UKHO.S100PermitService.API/Program.cs:118]: Please move all of these configuration strings into a private const string definition in Program.cs

File: src/UKHO.S100PermitService.API/Program.cs [hunk 8]
```diff
@@ -152,18 +152,28 @@ private static void ConfigureServices(WebApplicationBuilder builder)
                 client.Timeout = TimeSpan.FromMinutes(userPermitServiceApiConfiguration.RequestTimeoutInMinutes);
             });
 
+            var productKeyServiceApiConfiguration = builder.Configuration.GetSection("ProductKeyServiceApiConfiguration").Get<ProductKeyServiceApiConfiguration>();
```
Comment (jrippington) [inline src/UKHO.S100PermitService.API/Program.cs:155]: private const string definition

File: src/UKHO.S100PermitService.Common/Clients/IProductKeyServiceApiClient.cs [hunk 9]
```diff
@@ -0,0 +1,7 @@
+﻿namespace UKHO.S100PermitService.Common.Clients
+{
+    public interface IProductKeyServiceApiClient
+    {
+        Task<HttpResponseMessage> CallProductKeyServiceApiAsync(string uri, HttpMethod httpMethod, string payLoad, string accessToken, string correlationId);
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Clients/IProductKeyServiceApiClient.cs:5]: This method signature isn't right, it breaks encapsulation in all sorts of ways - please see comments on the client implementation.

File: src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    public class ProductKeyServiceApiClient : IProductKeyServiceApiClient
+    {
+        private readonly HttpClient _httpClient;
+
+        public ProductKeyServiceApiClient(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient();
+        }
+
+        public async Task<HttpResponseMessage> CallProductKeyServiceApiAsync(string uri, HttpMethod httpMethod, string payLoad, string accessToken, string correlationId)
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs:14]: The client should have one method per endpoint in the service implementation. The caller of this method should not have to construct the payload, specify the HTTP method - you shouldn't even care that it is calling a REST endpoint. Rename the method according to the endpoint on the PKS it is calling, and build the payload in this method from parameters passed in.

File: src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+            using var httpRequestMessage = new HttpRequestMessage(httpMethod, uri);
+            httpRequestMessage.Content = new StringContent(payLoad, Encoding.UTF8, "application/json");
+
+            if(!string.IsNullOrEmpty(accessToken))
+            {
+                httpRequestMessage.SetBearerToken(accessToken);
+                httpRequestMessage.AddHeader(PermitServiceConstants.XCorrelationIdHeaderKey, correlationId);
+            }
+
+            return await _httpClient.SendAsync(httpRequestMessage, CancellationToken.None);
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs:25]: You should specify a cancellation token here. 

The method should specify a cancellation token as a parameter, which should be passed here. The calling methods should all pass a cancellation token up to the root, where you should have a CancellationToken in the controller signature - this is bound to the HttpContext.RequestAborted property at runtime. Could you check for other instances and amend please?

File: src/UKHO.S100PermitService.Common/Services/PermitService.cs [hunk 12]
```diff
@@ -98,5 +107,15 @@ private static List<Products> GetProductsList()
             };
             return productsList;
         }
+
+        [ExcludeFromCodeCoverage]
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Services/PermitService.cs:111]: This method should definitely have unit tests, and should be called something like CreateProductKeyServiceRequests(...)

File: src/UKHO.S100PermitService.Common/Services/ProductKeyService.cs [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        /// <summary>
+        /// Get permit key from Product Key Service
+        /// </summary>
+        /// <param name="productKeyServiceRequest"></param>
+        /// <param name="correlationId"></param>
+        /// <returns>ProductKeyServiceResponse</returns>
+        /// <exception cref="Exception"></exception>
+        public async Task<List<ProductKeyServiceResponse>> PostProductKeyServiceRequestAsync(List<ProductKeyServiceRequest> productKeyServiceRequest, string correlationId)
```
Comment (jrippington) [inline src/UKHO.S100PermitService.Common/Services/ProductKeyService.cs:37]: The method name should tell a consumer what a method does, not how it does it - so should be something like GetPermitKeys(...). It should probably create the ProductKeyServiceRequest from parameters, not have it passed in, this breaks encapsulation.

**Review by jrippington — CHANGES_REQUESTED** [review]
Some refactoring needed, thanks!

File: src/UKHO.S100PermitService.API/Program.cs (hunk 7 above)
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.API/Program.cs]: Resolved

File: src/UKHO.S100PermitService.API/Program.cs (hunk 8 above)
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.API/Program.cs]: Resolved

File: src/UKHO.S100PermitService.Common/Clients/IProductKeyServiceApiClient.cs (hunk 9 above)
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.Common/Clients/IProductKeyServiceApiClient.cs]: Resolved

File: src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs (hunk 10 above)
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs]: Resolved

File: src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs (hunk 11 above)
Comment (MastekRohitChettri) [reply src/UKHO.S100PermitService.Common/Clients/ProductKeyServiceApiClient.cs]: Resolved

**Review by Mayuresh-Gawde — APPROVED** [review]
LGTM!!

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Review by Nadeem14456 — APPROVED** [review]

**Pull request merged**